server/src/main/java/org/elasticsearch/index/mapper/DynamicTemplate.java

```diff
@@ -301,30 +307,48 @@ static DynamicTemplate parse(String name, Map<String, Object> conf) throws Mappe
+        // match, path_match, and unmatch_mapping_type all imply
+        // "match_mapping_type: *" if not explicitly specified.
+        final boolean wildcardMatchMappingType = ((matchMappingType.isEmpty()
+            && matchPatternsAreDefined(match, pathMatch, unmatchMappingType))
+            || (matchMappingType.size() == 1 && matchMappingType.get(0).equals("*")));
+
+        Stream<XContentFieldType> matchXContentFieldTypes;
+        if (wildcardMatchMappingType) {
+            matchXContentFieldTypes = Stream.of(XContentFieldType.values());
+        } else {
             if (runtime) {
-                xContentFieldTypes = Arrays.stream(XContentFieldType.values())
-                    .filter(XContentFieldType::supportsRuntimeField)
-                    .toArray(XContentFieldType[]::new);
-            } else {
-                xContentFieldTypes = XContentFieldType.values();
-            }
-        } else if (matchMappingType != null) {
-            final XContentFieldType xContentFieldType = XContentFieldType.fromString(matchMappingType);
-            if (runtime && xContentFieldType.supportsRuntimeField() == false) {
-                throw new MapperParsingException(
-                    "Dynamic template ["
-                        + name
-                        + "] defines a runtime field but type ["
-                        + xContentFieldType
-                        + "] is not supported as runtime field"
-                );
+                final List<String> unsupported = matchMappingType.stream()
+                    .map(XContentFieldType::fromString)
+                    .filter(Predicate.not(XContentFieldType::supportsRuntimeField))
+                    .map(XContentFieldType::toString)
+                    .toList();
+                if (unsupported.isEmpty() == false) {
+                    final int numUnsupported = unsupported.size();
+                    throw new MapperParsingException(
+                        "Dynamic template ["
+                            + name
+                            + "] defines a runtime field but type"
+                            + (numUnsupported == 1 ? "" : "s")
+                            + " ["
+                            + String.join(", ", unsupported)
+                            + "] "
+                            + (numUnsupported == 1 ? "is" : "are")
+                            + " not supported as runtime field"
+                    );
+                }
             }
-            xContentFieldTypes = new XContentFieldType[] { xContentFieldType };
-        } else {
-            xContentFieldTypes = new XContentFieldType[0];
+            matchXContentFieldTypes = matchMappingType.stream().map(XContentFieldType::fromString);
         }
+        if (runtime) {
+            matchXContentFieldTypes = matchXContentFieldTypes.filter(XContentFieldType::supportsRuntimeField);
+        }
```

Comment:
I wonder if we should move this into the `if (wildcardMatchMappingType)` branch as the `else` branch is already guaranteed to not contain runtime-unsupported field types if `runtime` is `true`. I felt like it took me another couple of seconds to understand that when reading the code.

Reply:
Good pick up, that's a nice simplification.